llm/kv_cache.go

```diff
@@ -0,0 +1,102 @@
+type GGMLModel interface {
+	KV() KV
+}
+
+// modelSupportsFlashAttention checks if the model supports flash attention
+func modelSupportsFlashAttention(ggml GGMLModel) bool {
+	// Check if it's an embedding model - embedding models don't support flash attention
+	if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
+		return false
+	}
+
+	// Check head counts match and are non-zero
+	headCountK := ggml.KV().EmbeddingHeadCountK()
+	headCountV := ggml.KV().EmbeddingHeadCountV()
+	return headCountK != 0 && headCountV != 0 && headCountK == headCountV
+}
+
+// ValidKVCacheTypes contains all supported KV cache types
+// "q5_1", "q5_0", "iq4_nl", "q4_1" are also supported by llama.cpp, we're just not enabling them in Ollama
+var ValidKVCacheTypes = []string{"f32", "f16", "q8_0", "q4_0"}
+
+// kVCacheQuantization determines the appropriate KV cache quantization type based on
+// model characteristics, GPU capabilities, and environment settings
+func kVCacheQuantization(ggml GGMLModel, gpus discover.GpuInfoList) string {
+	kvct := envconfig.KvCacheType()
+	if kvct == "" {
+		return "f16" // default value
+	}
+
+	// For embedding models, only allow f16 and f32
+	if !modelSupportsFlashAttention(ggml) {
+		if kvct == "f32" {
+			return "f32"
+		}
+		if kvct != "f16" {
+			slog.Warn("only f16 and f32 cache types are supported for embedding models, defaulting to f16",
+				"type", kvct)
+		}
+		return "f16"
+	}
+
+	// Check flash attention support
+	flashAttnRequested := envconfig.FlashAttention()
+	if !flashAttnRequested || !gpus.SupportsFlashAttention() || !modelSupportsFlashAttention(ggml) {
+		// Without flash attention, default to f16
+		if flashAttnRequested {
+			slog.Warn("flash attention requested, but not supported by model or hardware, defaulting to f16")
+		}
+		return "f16"
+	}
+
+	// Validate against supported types
+	if !slices.Contains(ValidKVCacheTypes, kvct) {
+		slog.Warn("invalid cache type, defaulting to f16", "type", kvct)
+		return "f16"
+	}
+
+	return kvct
+}
+
+// estimateKvCacheSize determines the total memory required for KV cache based on the quantization type
+func estimateKvCacheSize(cacheType string, numCtx uint64, ggml *GGML) uint64 {
+	var bytesPerElement float64
+	isEmbeddingModel := !modelSupportsFlashAttention(ggml)
+	kv := ggml.KV()
+
+	// For embedding models, automatically convert to f16
+	if isEmbeddingModel && cacheType != "f32" {
+		cacheType = "f16"
+	}
+
+	switch cacheType {
+	case "f32", "fp32":
+		bytesPerElement = 4 // fp32
+	case "", "f16", "fp16":
+		bytesPerElement = 2 // fp16
+	case "q8_0":
+		bytesPerElement = 1 // 1/2 of fp16
+	case "q4_0":
+		bytesPerElement = 0.5 // 1/4 of fp16
+	default:
+		// Default to fp16 if unknown
+		bytesPerElement = 2
+		slog.Warn("Unknown cache type, defaulting to fp16", "type", cacheType)
+	}
+
+	estimate := uint64(float64(numCtx*kv.BlockCount()*kv.EmbeddingHeadCountK()*kv.HeadCountKV()) * bytesPerElement)
```

Comment:
```suggestion
	estimate := uint64(float64(numCtx*kv.BlockCount()*(kv.EmbeddingHeadCountK()+kv.EmbeddingHeadCountV())*kv.HeadCountKV()) * bytesPerElement)
```

and then drop the `* 2` a few lines down.